File: 2015/python/day01.py
#!/usr/bin/env python3
import sys


def problem1(s: str) -> int:
    ret = 0
    for c in s:
        if c == "(":
            ret += 1
        else:
            ret -= 1
    return ret


def problem2(s: str) -> int:
    pos = 0
    for i, c in enumerate(s):
        if c == "(":
            pos += 1
        else:
            pos -= 1

        if pos == -1:
            return i + 1

    raise Exception("never reach here")


def test():
    def check(s: str, got: int, expected: int):
        try:
            assert got == expected
        except AssertionError:
            print(f"input={s} got={got}, expected={expected}")
            sys.exit(1)

    test_data1 = [
        ("(())", 0),
        ("()()", 0),
        ("(((", 3),
        ("(()(()(", 3),
        ("))(((((", 3),
        ("())", -1),
        ("))(", -1),
        (")))", -3),
        (")())())", -3),
    ]
    for s, expected in test_data1:
        check(s, problem1(s), expected)

    test_data2 = [(")", 1), ("()())", 5)]
    for s, expected in test_data2:
        check(s, problem2(s), expected)


test()


with open("../input/day01.txt") as f:
    s = f.readline()
    ret1 = problem1(s)
    print(f"ret1={ret1} expected=232")

    ret2 = problem2(s)
    print(f"ret2={ret2}, expected=1783")

File: 2019/cpp/day01.cpp
#include <cassert>
#include <vector>
#include <iostream>
#include <print>

auto get_fuel(int n) -> int {
    return (n / 3) - 2;
}

auto problem1(const std::vector<int> &nums) -> int {
    int ret = 0;
    for (int num : nums) {
        ret += get_fuel(num);
    }

    return ret;
}

auto get_fuel2(int n) -> int {
    int ret = 0;
    while (n > 0) {
        int m = (n / 3) - 2;
        if (m <= 0) {
            break;
        }

        ret += m;
        n = m;
    }

    return ret;
}

auto problem2(const std::vector<int> &nums) -> int {
    int ret = 0;
    for (int num : nums) {
        ret += get_fuel2(num);
    }

    return ret;
}

void test() {
    assert(get_fuel(12) == 2);
    assert(get_fuel(14) == 2);
    assert(get_fuel(1969) == 654);
    assert(get_fuel(100756) == 33583);

    assert(get_fuel2(14) == 2);
    assert(get_fuel2(1969) == 966);
    assert(get_fuel2(100756) == 50346);
}

auto main() -> int {
    test();

    std::vector<int> input;
    int n;
    while (std::cin >> n) {
        input.push_back(n);
    }

    int ret1 = problem1(input);
    std::println("problem1={}", ret1);

    int ret2 = problem2(input);
    std::println("problem2={}", ret2);
    return 0;
}
